File: player/src/main/java/com/xyoye/player/subtitle/SubtitleUtils.java
package com.xyoye.player.subtitle;

/**
 * Created by XYJ on 2020/2/13.
 */

public class SubtitleUtils {
    private static boolean isTopSubtitle = false;
    private static boolean isEmptySubtitle = false;

    public static void showSubtitle(SubtitleView subtitleView, String subtitle, int textSize) {
        //字幕分行
        SubtitleView.SubtitleText[] subtitleTexts;
        if (subtitle.contains("\\N")) {
            subtitleTexts = getSubtitleTexts(textSize, subtitle.split("\\N{ACKNOWLEDGE}"));
        } else if (subtitle.contains("\n")) {
            subtitleTexts = getSubtitleTexts(textSize, subtitle.split("\n"));
        } else {
            subtitleTexts = getSubtitleTexts(textSize, subtitle);
        }

        if (isEmptySubtitle){;
            subtitleView.setTopTexts(subtitleTexts);
            subtitleView.setBottomTexts(subtitleTexts);
            isEmptySubtitle = false;
            return;
        }

        if (isTopSubtitle){
            subtitleView.setTopTexts(subtitleTexts);
        } else {
            subtitleView.setBottomTexts(subtitleTexts);
        }
    }

    private static SubtitleView.SubtitleText[] getSubtitleTexts(int textSize, String... subtitles) {
        isEmptySubtitle = false;
        //字幕为空，清除上一次的字幕
        if (subtitles.length == 1 && subtitles[0].length() == 0){
            SubtitleView.SubtitleText[] subtitleTexts = new SubtitleView.SubtitleText[1];
            subtitleTexts[0] = new SubtitleView.SubtitleText();
            isEmptySubtitle = true;
            return subtitleTexts;
        }

        isTopSubtitle = false;
        SubtitleView.SubtitleText[] subtitleTexts = new SubtitleView.SubtitleText[subtitles.length];
        for (int i = 0; i < subtitles.length; i++) {
            String subtitle = subtitles[i].trim();
            SubtitleView.SubtitleText subtitleText = new SubtitleView.SubtitleText();
            subtitleText.setSize(textSize);
            //第一行以{开头，则认为是特殊字幕，现显示在顶部
            if (subtitle.startsWith("{")){
                if (i == 0){
                    isTopSubtitle = true;
                }
                //忽略{}中内容
                int endIndex = subtitle.lastIndexOf("}") + 1;
                if (endIndex != 0 && endIndex <= subtitle.length()){
                    subtitleText.setText(subtitle.substring(endIndex));
                } else {
                    subtitleText.setText(subtitle);
                }
            } else {
                //普通内容显示在底部
                isTopSubtitle = false;
                subtitleText.setText(subtitle);
            }

            subtitleTexts[i] = subtitleText;
        }
        return subtitleTexts;
    }
}

File: player_component/src/main/java/com/xyoye/subtitle/SubtitleUtils.kt
package com.xyoye.subtitle

import android.graphics.Color
import com.xyoye.subtitle.info.Caption

/**
 * Created by xyoye on 2020/12/15.
 */

object SubtitleUtils {

    /**
     * 文字转换显示需要字幕格式
     */
    fun caption2Subtitle(captionText: String?): MutableList<SubtitleText> {
        if (captionText.isNullOrEmpty())
            return mutableListOf()

        val newCaption = Caption()
        newCaption.content = captionText
        return caption2Subtitle(newCaption)
    }

    /**
     * caption转换显示需要字幕格式
     */
    fun caption2Subtitle(caption: Caption): MutableList<SubtitleText> {
        //字幕颜色
        val subtitleColor = getCaptionColor(caption.style?.color)

        //分割每行字幕
        val subtitle = caption.content.replace("<br />", "\n")
        val upperRegex = "\\N"
        val lowerRegex = "\n"
        if (subtitle.contains(upperRegex)) {
            return strings2Subtitle(subtitleColor, *(subtitle.split(upperRegex).toTypedArray()))
        }

        if (subtitle.contains(lowerRegex)) {
            return strings2Subtitle(subtitleColor, *(subtitle.split(lowerRegex).toTypedArray()))
        }

        return strings2Subtitle(subtitleColor, subtitle)
    }

    private fun strings2Subtitle(
        subtitleColor: Int,
        vararg subtitles: String
    ): MutableList<SubtitleText> {
        val subtitleList = mutableListOf<SubtitleText>()

        for (subtitle in subtitles) {
            //第一行以{开头，则认为是特殊字幕，现显示在顶部
            if (subtitle.startsWith("{")) {
                val endIndex = subtitle.lastIndexOf("}") + 1
                subtitleList.add(
                    if (endIndex != 0 && endIndex <= subtitle.length) {
                        //忽略{}中内容
                        SubtitleText(
                            subtitle.substring(
                                endIndex
                            ), true, subtitleColor
                        )
                    } else {
                        SubtitleText(
                            subtitle,
                            true,
                            subtitleColor
                        )
                    }
                )
            } else {
                subtitleList.add(
                    SubtitleText(
                        subtitle,
                        false,
                        subtitleColor
                    )
                )
            }
        }

        return subtitleList
    }


    /**
     * 获取字幕颜色
     */
    private fun getCaptionColor(colorStr: String?): Int {
        var rgbaText = colorStr

        if (rgbaText.isNullOrEmpty())
            return Color.WHITE

        if (!rgbaText.startsWith("#"))
            rgbaText = "#$colorStr"

        //颜色字符串为rgba格式，要转换成argb
        return try {
            val rgba = Color.parseColor(rgbaText)
            rgba ushr 8 or (rgba shl 32 - 8)
        } catch (e: IllegalArgumentException) {
            Color.WHITE
        }
    }
}